### PR TITLE
Implement basic App foundation

### LINKED ISSUE #8: [Basic] Save contact data locally
As a programmer, the contact data I enter will be saved locally on my hard drive.

### PR DESCRIPTION
**Pull request by marcusbory**
Fixes #8 #9 

- App now has local storage feature implemented
- Saving can also be done
- Happy flow is ok
- No error handling (app is fragile, but can be built upon by the rest of you)
- <strike>Add contact will need to be improved upon to have randomly ordered tags</strike>
- <strike>Add contact will require (a lot) more exceptions/error handling</strike>

UPDATE 6/10/21 thanks to @ng-andre, I will be reimplementing the add feature with separate parsers. This PR can be merged to fix #8 and #9 first. KIV @lezongmun

### CODE REVIEW

**Review by lezongmun — COMMENTED** [review]
Code is great! But there are some checks that have failed. There seems to have some problems with the CheckStyle and the IO Redirection. Perhaps you can run check style scan on your code to find the errors and push a fix to this PR. Also, remember to update the input.txt and expected.txt as well for the IO Redirection tests.

**Comment by marcusbory** [comment]
Hi @lezongmun , appreciate your help and guidance in reviewing my PR. I have made changes to the IO redirection tests, and similarly added the gitignore files as follows so that, going forward, all future tests will start from a fresh build and fresh local storage. Do go through my updated PR and let me know if merging is possible. I will be more than willing to explain any uncertain parts you may have.

File: src/main/java/seedu/storage/ContactsDecoder.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import seedu.contact.ContactList;
+import seedu.contact.DetailType;
+import seedu.exception.FileErrorException;
+import static seedu.storage.Storage.SEPARATOR;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.util.Scanner;
+
+public class ContactsDecoder {
```
Comment (lezongmun) [inline src/main/java/seedu/storage/ContactsDecoder.java:13]: I like the use of ContactsDecoder and ContactsEncoder to abstract the code, makes it a lot more readable.

File: src/main/java/seedu/contact/DetailType.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package seedu.contact;
+
+/**
+ * Enumeration for Contact Detail Type.
+ * This is used for decoding contacts from CSV. The indexes of CSV file lines
+ * upon splitting will be stored here.
+ * Eg. ANDRE, ng-andre => upon splitting becomes ["ANDRE", "ng-andre"]
```
Comment (lezongmun) [inline src/main/java/seedu/contact/DetailType.java:7]: Great use of examples and comments to show how the code is expected to work.

**Review by lezongmun — APPROVED** [review]
Overall great use of abstractions and comments to make the code easy to understand. It is noted that only #8 and #9 has been implemented in this PR, but other than that, LGTM!

**Comment by lezongmun** [comment]
> Hi @lezongmun , appreciate your help and guidance in reviewing my PR. I have made changes to the IO redirection tests, and similarly added the gitignore files as follows so that, going forward, all future tests will start from a fresh build and fresh local storage. Do go through my updated PR and let me know if merging is possible. I will be more than willing to explain any uncertain parts you may have.
> 
> In addition, I have also made sure that my code was compliant with the CheckStyle.

Hello @marcusbory, I have looked through the updated code, everything looks great now. I have also noted the changes to the IO redirections tests and gitignore files as well as the rationale behind them. Thanks Marcus!

File: src/main/java/seedu/parser/Parser.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
+package seedu.parser;
+
+import seedu.command.AddContactCommand;
+import seedu.command.Command;
+import seedu.command.ExitCommand;
+import seedu.command.FailedCommand;
+
+public class Parser {
+    private static final String ADD_TASK_COMD = "add";
```
Comment (ng-andre) [inline src/main/java/seedu/parser/Parser.java:9]: @marcusbory this constant should be ADD_CONTACT_COMD

**Pull request merged**